File: js/ABC13/8/abc138_d.js
// D - Ki
// TLE
'use strict'

function main(inp) {
  inp = inp.trim().split('\n')
  const NQ = inp.shift().split(' ').map(n => Number(n))
  let ans = new Array(NQ[0]+1).fill(0)
  let ab = {}
  let px = {}

  for (let i = 0; i < inp.length; i++) {
    const arr = inp[i].split(' ').map(n => Number(n))
    let j = arr[0]
    if (i < NQ[0] - 1) {
      if(!(j in ab)) ab[j] = []
      ab[j].push(arr[1])
    } else {
      if(!(j in px)) px[j] = 0
      px[j] += arr[1]
    }
  }

  let stack = [1]
  const countup = n => {
    if (n in px) ans[n] += px[n]
    if (n in ab) {
      for (let i = 0; i < ab[n].length; i++) {
        let j = ab[n][i]
        ans[j] += ans[n]
        stack.push(j)
      }
    }
  }
  while(stack.length) {
    countup(stack.pop())
  }
  ans.shift()

  console.log(ans.join(' '))
}

main(`6 2
1 2
1 3
2 4
3 6
2 5
1 10
1 10`)

File: js/ABC13/8/abc138_d.py
## D - Ki
## TLE

N, Q = map(int, input().split())
ab = {}
pq = {}
for i in range(N-1):
  arr = input().split()
  if not arr[0] in ab:
    ab[arr[0]] = []
  ab[arr[0]].append(arr[1])
for i in range(Q):
  arr = input().split()
  if not arr[0] in pq:
    pq[arr[0]] = 0
  pq[arr[0]] += int(arr[1])

ans = [0] * (N + 1)
que = ['1']
while len(que) > 0:
  n = str(que.pop(0))
  if n in pq:
    ans[int(n)] += int(pq[n])
  if n in ab:
    for i in range(len(ab[n])):
      j = int(ab[n][i])
      ans[j] += ans[int(n)]
      que.append(str(j))
ans.pop(0)

print(' '.join(map(str, ans)))